### PR TITLE
Feature/ai 136/sob sql tool

### PR DESCRIPTION
**Pull request by simdage**

### CODE REVIEW

File: _scripts/evaluate_tools.py [hunk 1]
```diff
@@ -59,27 +59,31 @@ def get_actual_results(langchain_tracer: LangChainTracer, name: str) -> Tuple[Di
 
     if args.eval_path != "None":
         # getting eval df
-        eval_df = pd.read_csv(args.eval_path, sep=";")
+        eval_df = pd.read_csv(args.eval_path, sep=";").iloc[:2, :]
```
Comment (dtourillon-cropguard) [inline _scripts/evaluate_tools.py]: Aren't we just reading the first example with this change? If so, this is really what we want or just debug/temporary changes?

File: main.py [hunk 2]
```diff
@@ -14,8 +14,8 @@
 
 from pydantic.v1 import BaseModel
 
-from croptalk.model_llm import model
-# from croptalk.model_openai_functions import model
+# from croptalk.model_llm import model
+from croptalk.model_openai_functions import model
```
Comment (dtourillon-cropguard) [inline main.py]: That change should already be in `openai_functions` branch by now, so you just need to pull changes from it

File: _scripts/evaluate_tools.py (hunk 1 above)
Comment (simdage) [reply _scripts/evaluate_tools.py]: I was testing the pipeline and wanted to save on credits.. will revert :)

File: croptalk/tools.py [hunk 3]
```diff
@@ -158,10 +271,135 @@ def get_wfrp_commodities(reinsurance_yr: str, state_code: str, county_code: str)
...
+
+full_prompt = ChatPromptTemplate.from_messages(
+    [
+        SystemMessagePromptTemplate(prompt=few_shot_prompt),
+        ("human", "{input}"),
+        MessagesPlaceholder("agent_scratchpad"),
+    ]
+)
+
+TOOLS = [get_sob_metrics_sql_agent, get_wfrp_commodities, get_sp_document]
```
Comment (dtourillon-cropguard) [inline croptalk/tools.py:369]: I noticed that validate_query is not in TOOLS... I think it's on purpose, but just wanted to validate with you...?

File: croptalk/tools.py [hunk 4]
```diff
@@ -125,6 +150,94 @@ def get_sp_document(state: Optional[str] = None,
...
+class SQLHandler(BaseCallbackHandler):
+    def __init__(self):
+        self.sql_result = []
+
+    def on_agent_action(self, action, **kwargs):
+        """Run on agent action. if the tool being used is sql_db_query,
+         it means we're submitting the sql and we can
+         record it as the final sql"""
+
+        if action.tool in ["sql_db_query_checker", "sql_db_query"]:
```
Comment (dtourillon-cropguard) [inline croptalk/tools.py:162]: out of curiosity, what are these 2 names matching exactly? I was expecting them to match tool names we created but it doesn't appear to be the case...

**Review by dtourillon-cropguard — APPROVED** [review]
I provided a few minor comments, otherwise LGTM

File: croptalk/tools.py (hunk 3 above)
Comment (simdage) [reply croptalk/tools.py:369]: yes, I integrated it directly with the sql agent

File: croptalk/tools.py (hunk 4 above)
Comment (simdage) [reply croptalk/tools.py:162]: this is used to match internal tools used by the SQLAgentExecutor. It's only meant to return the exact query used by the agent and not any tools we developped

**Pull request merged**